zerver/lib/message.py

```diff
@@ -344,3 +344,169 @@ def render_markdown(message, content, realm=None, realm_alert_words=None, messag
         message.is_me_message = Message.is_status_message(content, rendered_content)
 
     return rendered_content
+
+def huddle_users(recipient_id):
+    # type: (int) -> str
+    display_recipient = get_display_recipient_by_id(recipient_id,
+                                                    Recipient.HUDDLE,
+                                                    None)  # type: ignore
+    user_ids = [obj['id'] for obj in display_recipient]  # type: ignore
+    user_ids = sorted(user_ids)
+    return ','.join(str(uid) for uid in user_ids)
+
+def aggregate_dict(input_rows, lookup_fields, input_field, output_field):
+    # type: (List[Dict[str, Any]], List[str], str, str) -> List[Dict[str, Any]]
+    lookup_dict = dict()  # type: Dict[Any, Dict]
+
+    for input_row in input_rows:
+        lookup_key = tuple([input_row[f] for f in lookup_fields])
+        if lookup_key not in lookup_dict:
+            obj = {}
+            for f in lookup_fields:
+                obj[f] = input_row[f]
+            obj[output_field] = []
+            lookup_dict[lookup_key] = obj
+
+        lookup_dict[lookup_key][output_field].append(input_row[input_field])
+
+    sorted_keys = sorted(lookup_dict.keys())
+
+    return [lookup_dict[k] for k in sorted_keys]
+
+def get_unread_message_ids_per_recipient(user_profile):
+    # type: (UserProfile) -> Dict[str, List[Dict[str, Any]]]
+    user_msgs = UserMessage.objects.filter(
+        user_profile=user_profile
+    ).extra(
+        where=[UserMessage.where_unread()]
+    ).values(
+        'message_id',
+        'message__sender_id',
+        'message__subject',
+        'message__recipient_id',
+        'message__recipient__type',
+        'message__recipient__type_id',
+    )
+
+    rows = list(user_msgs)
+
+    pm_msgs = [
+        dict(
+            sender_id=row['message__sender_id'],
+            message_id=row['message_id'],
+        ) for row in rows
+        if row['message__recipient__type'] == Recipient.PERSONAL]
+
+    pm_objects = aggregate_dict(
+        input_rows=pm_msgs,
+        lookup_fields=[
+            'sender_id',
+        ],
+        input_field='message_id',
+        output_field='unread_message_ids',
+    )
+
+    stream_msgs = [
+        dict(
+            stream_id=row['message__recipient__type_id'],
+            topic=row['message__subject'],
+            message_id=row['message_id'],
+        ) for row in rows
+        if row['message__recipient__type'] == Recipient.STREAM]
+
+    stream_objects = aggregate_dict(
+        input_rows=stream_msgs,
+        lookup_fields=[
+            'stream_id',
+            'topic',
+        ],
+        input_field='message_id',
+        output_field='unread_message_ids',
+    )
+
+    huddle_msgs = [
+        dict(
+            recipient_id=row['message__recipient_id'],
+            message_id=row['message_id'],
+        ) for row in rows
+        if row['message__recipient__type'] == Recipient.HUDDLE]
```

Comment:
would it be a good idea to extract this list generation as a `get_recipient_msgs` function inside this function accepting `rows` and `Recipient.HUDDLE` or `Recipient.STREAM` or `Recipient.PERSONAL`?
We can de duplicate this list generation with all three (streams, personal, huddles).